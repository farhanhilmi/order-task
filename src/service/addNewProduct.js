import productRepo from '../repository/productRepo.js';
import { validateProduct } from '../utils/validation.js';

const addNewProduct = async (data) => {
  try {
    const { error } = validateProduct(data);
    if (error) {
      throw new Error(error.details.map((err) => err.message));
    }
    return await productRepo.addProduct(data);
  } catch (err) {
    return err;
  }
};

export default addNewProduct;
