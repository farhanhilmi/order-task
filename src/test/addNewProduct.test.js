import addNewProduct from '../service/addNewProduct.js';

describe('Product Service | Add New Product', () => {
  it('should store a new product with correct property', async () => {
    const newProduct = {
      name: 'apple',
      quantity: 50,
      description: 'ladkald',
      price: 10000,
    };
    const product = await addNewProduct(newProduct);
    expect(product).toMatchSnapshot();
    expect(product.name).toBeTruthy();
    expect(product.quantity).toBeTruthy();
    expect(product.description).toBeTruthy();
    expect(product.price).toBeTruthy();
  });

  it('should return an error message due to missing field', async () => {
    const newProduct = {
      name: 'apple',
      quantity: 50,
      // description: 'ladkald',
      price: 10000,
    };
    const products = await addNewProduct(newProduct);
    const throwError = () => {
      throw new TypeError(products);
    };
    expect(throwError).toThrow(`Error: description is a required field`);
  });

  it('should return an error message due to wrong input type', async () => {
    const newProduct = {
      name: 'apple',
      quantity: '50k',
      description: 'ladkald',
      price: 10000,
    };
    const products = await addNewProduct(newProduct);
    const throwError = () => {
      throw new TypeError(products);
    };
    expect(throwError).toThrow(`Error: quantity should be a type of 'number'`);
  });
});
