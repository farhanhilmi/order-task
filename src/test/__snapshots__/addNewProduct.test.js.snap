// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Service | Add New Product should store a new product with correct property 1`] = `
Object {
  "__v": 0,
  "_id": "62903f14e8db779d6abe5607",
  "createdDate": 2022-05-27T03:01:40.383Z,
  "description": "ladkald",
  "modifyDate": 2022-05-27T03:01:40.383Z,
  "name": "apple",
  "price": 10000,
  "quantity": 50,
}
`;
